core/trino-main/src/main/java/io/trino/operator/output/columnar/PositionsAppenderFactory.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output.columnar;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import io.airlift.bytecode.DynamicClassLoader;
+import io.trino.collect.cache.NonEvictableLoadingCache;
+import io.trino.operator.output.columnar.PositionsAppender.TypedPositionsAppender;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.Int128ArrayBlock;
+import io.trino.spi.block.Int96ArrayBlock;
+import io.trino.spi.type.FixedWidthType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VariableWidthType;
+import io.trino.sql.gen.IsolatedClass;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.Objects;
+
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCache;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Isolates the {@code PositionsAppender} class per type and block tuples.
+ * Type specific {@code PositionsAppender} implementations manually inline {@code Type#appendTo} method inside the loop
+ * to avoid virtual(mega-morphic) calls and force jit to inline the {@code Block} and {@code BlockBuilder} methods.
+ */
+public class PositionsAppenderFactory
+{
+    private final NonEvictableLoadingCache<CacheKey, PositionsAppender> cache;
+
+    public PositionsAppenderFactory()
+    {
+        this.cache = buildNonEvictableCache(
+                CacheBuilder.newBuilder().maximumSize(1000),
+                CacheLoader.from(key -> createAppender(key.type)));
+    }
+
+    public PositionsAppender create(Type type, Class<? extends Block> blockClass)
+    {
+        return cache.getUnchecked(new CacheKey(type, blockClass));
+    }
+
+    private PositionsAppender createAppender(Type type)
+    {
+        Class<? extends PositionsAppender> appenderClass = findDedicatedAppenderClassFor(type);
+
+        if (appenderClass == null) {
+            return new TypedPositionsAppender(type);
```

Comment:
remaind me again why we couldn't just isolate this guy (`type.appendTo` intuitively should be inlined since it's called from single isolated callsite). Please add comment in code.